### ISSUE TITLE
Add Cox proportional hazards model

### ISSUE THREAD

**Issue by dustinvtran** [airoldilab/sgd#23]
Experiment to run them, stock data set for them, and compare with `coxnet`

**Comment by dustinvtran** [comment]
Added in beee5eb5c7ff46ea270b8d831defdd4fd292b508

**Comment by MarcinKosinski** [comment]
Hi,

I really appreciate the huge work you have done while developing this package. Great work I have to say. I found this package by typing `stochastic gradient descent estimation in cox model` in google websearch. In `README.md` file I've spotted that you mentioned [https://github.com/airoldilab/sgd/edit/master/README.md#L55](https://github.com/airoldilab/sgd/edit/master/README.md#L55) that cox loss function for proportional hazard is implemented. I am really curious about that!

As far as I looked around at your documentation and source code of this package I've noticed that there is some kind of implementation of [cox parameters estimation in `C`](https://github.com/airoldilab/sgd/blob/master/src/model/cox_model.h) (but unfortunately there is still a line with a comment [TODO](https://github.com/airoldilab/sgd/blob/master/src/model/cox_model.h#L41) - is it finished?) and you can specify a `model` parameter in `sgd()` function which can be set to `lm` or `glm` (should I suppose that cox proportional harads model is a kind of generalized linear model?).

*I have a big question.* Did you implement a stochastic gradient descent method of estimation parameters in the cox proportional hazards model? I am not so skilled in `C` (I'm rather an `R` enthusiast`). Could you please upload a link for referencing article where people proof a form and algorithm for stochastic gradient descent estimation in cox model ph model (I can't find a single paper about that)? 

How should I specify a `sgd()` function to run stochastic gradient descent estimation for cox ph model? There is not a single word about `cox ph` in `sgd`documentation and you have really few examples, so I can't figure out how to use this function to get what I wanted :) Maybe you should prepare a `knitr` vignette with most popular and valuable use-cases of `sgd` package?

Thanks for response,
Hope to hear from you soon,
Kind regards,

Marcin

**Comment by ptoulis** [comment]
Hi Marcin

first, thank you very much for your interest.

please find my response below.
Dustin, the lead developer, can jump in if I missed something in my
response.


On Wed, Sep 2, 2015 at 5:01 PM, Marcin Kosiński <notifications@github.com>
wrote:

> Hi,
>
> I really appreciate the huge work you have done while developing this
> package. Great work I have to say. I found this package by typing stochastic
> gradient descent estimation in cox model in google websearch. In README.md
> file I've spotted that you mentioned
> https://github.com/airoldilab/sgd/edit/master/README.md#L55 that cox loss
> function for proportional hazard is implemented. I am really curious about
> that!
>
> As far as I looked around at your documentation and source code of this
> package I've noticed that there is some kind of implementation of cox
> parameters estimation in C
> <https://github.com/airoldilab/sgd/blob/master/src/model/cox_model.h>
> (but unfortunately there is still a line with a comment TODO
> <https://github.com/airoldilab/sgd/blob/master/src/model/cox_model.h#L41>
> - is it finished?) and you can specify a model parameter in sgd()
> function which can be set to lm or glm (should I suppose that cox
> proportional harads model is a kind of generalized linear model?).
>

We are very close to an implementation. There is a working version but it
is not complete yet.
Main problems arise from the non-convexities of the Cox model loss function.
Technically the Cox model is not a generalized linear model, but it
resembles such models because it depends
on the model parameters through a linear combination with covariates (i.e.,
the terms exp(X'θ))

> *I have a big question.* Did you implement a stochastic gradient descent
> method of estimation parameters in the cox proportional hazards model? I am
> not so skilled in C (I'm rather an R enthusiast`). Could you please
> upload a link for referencing article where people proof a form and
> algorithm for stochastic gradient descent estimation in cox model ph model
> (I can't find a single paper about that)?
>
We have implemented SGD for the Cox model, in the sense that we have a
version that works most of the times and estimates
the true parameters. But the implementation is not 100% ready yet.
Also there is no theory of SGD + Cox to the best of our knowledge. This was
one our motivations
for adding this model to our package.

> How should I specify a sgd() function to run stochastic gradient descent
> estimation for cox ph model? There is not a single word about cox ph in
> sgddocumentation and you have really few examples, so I can't figure out
> how to use this function to get what I wanted :) Maybe you should prepare a
> knitr vignette with most popular and valuable use-cases of sgd package?
>

Because our SGD for Cox model is not ready yet, the option is not
implemented in the main package.
We still have work to do to finish the code to our satisfaction, and then
create the proper documentation
(writing a knitr entry is great -- I added it as an issue)

But this is a good opportunity for us.
Can I ask what kind of application you consider where SGD + Cox is needed?
It would be helpful for us to work on a concrete example if the data are
publicly available.



> Thanks for response,
> Hope to hear from you soon,
> Kind regards,
>
> Marcin
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/airoldilab/sgd/issues/23#issuecomment-137242996>.
>



-- 
best
panos

**Comment by dustinvtran** [comment]
Hi @MarcinKosinski, you'll have to download the latest development version from Github in order to use it:
```{R}
if (packageVersion("devtools") < 1.6) {
  install.packages("devtools")
}
devtools::install_github("airoldilab/sgd")
```
You can then check out an example in `demo/cox-regression.R`. @ptoulis is spot on about the lacking theory. We're currently developing it, and any comments on where you're interested in using large-scale cox models would go a long way!

**Comment by MarcinKosinski** [comment]
Thanks for response!

I am curious about that because I am writing a master thesis about sgd estimation in cox ph model so nothing special :). I was wondering if someone has already developed theory or implemented an R package about that. I've seen that in cox C code you wrote a comment that you assume that observations are arranged by date - why this is important assumption?

**Comment by ptoulis** [comment]
Hi Marcin

there are several definitions for the Cox model.

We are using one of straightforward one where the data are event times
(either something happened or censoring)
Without loss of generality these can be assumed to be ordered.

Check out, for example, Ch.10, Davison's Statistical Models, 2003.

On Thu, Sep 3, 2015 at 4:53 AM, Marcin Kosiński <notifications@github.com>
wrote:

> Thanks for response!
>
> I am curious about that because I am writing a master thesis about sgd
> estimation in cox ph model so nothing special :). I was wondering if
> someone has already developed theory or implemented an R package about
> that. I've seen that in cox C code you wrote a comment that you assume that
> observations are arranged by date - why this is important assumption?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/airoldilab/sgd/issues/23#issuecomment-137381843>.
>



-- 
best
panos

**Comment by ptoulis** [comment]
By the way, if there is still time for your thesis, we will soon be able to share a short document with 
our implementation of SGD for Cox models :)

**Comment by MarcinKosinski** [comment]
I thought there is only 1 definition of the Cox proportional hazards
model, stated by Cox himself in his work: `Regression Models and
Life-Tables, D. R. Cox Journal of the Royal Statistical Society. Series B
(Methodological), Vol. 34, No. 2. (1972), pp. 187-220.` Although it was
defined for life-tables it can be easily generalized for single
observations.

I'll check that Ch.10, Davison's Statistical Models, 2003 - thanks for
bibliography.

You wrote previously that

>
> Main problems arise from the non-convexities of the Cox model loss
function.
>

What do you mean when you say `loss function`? I thought that inference and
estimation in that model should be based on partial-(log)-likelihood, which
for right censored data where censoring is non-informative and independent
is in fact a full (log)-likelihood. But I might be wrong :)

Good to hear that someone skilled is working on the problem of large-scale
estimation based on stochastic gradient descent for the cox proportional
hazards model!
I am planning to finish my thesis at the end of this month and It would be
great to cite your work and try to use your implementation. But if that
time Harvard's Data Science
Lab won't publish anything I'm afraid I'll need to invent this
implementation by myself (in which I have less faith :)).



2015-09-03 21:46 GMT+02:00 Panos Toulis <notifications@github.com>:

**Comment by ptoulis** [comment]
On Fri, Sep 4, 2015 at 5:16 AM, Marcin Kosiński <notifications@github.com>
wrote:

> I thought there is only 1 definition of the Cox proportional hazards
> model, stated by Cox himself in his work: `Regression Models and
> Life-Tables, D. R. Cox Journal of the Royal Statistical Society. Series B
> (Methodological), Vol. 34, No. 2. (1972), pp. 187-220.` Although it was
> defined for life-tables it can be easily generalized for single
> observations.
>
>
there are several variations by now; for example, models with
time-dependence.


> I'll check that Ch.10, Davison's Statistical Models, 2003 - thanks for
> bibliography.
>
> You wrote previously that
>
> >
> > Main problems arise from the non-convexities of the Cox model loss
> function.
> >
>
> What do you mean when you say `loss function`? I thought that inference and
> estimation in that model should be based on partial-(log)-likelihood, which
> for right censored data where censoring is non-informative and independent
> is in fact a full (log)-likelihood. But I might be wrong :)
>

negative log likelihood = loss, so we are talking about the same thing :)
the partial likelihood has this sigmoid terms which can be non-convex.
for example, similar problems can arise in logistic regression.



>
> Good to hear that someone skilled is working on the problem of large-scale
> estimation based on stochastic gradient descent for the cox proportional
> hazards model!
> I am planning to finish my thesis at the end of this month and It would be
> great to cite your work and try to use your implementation. But if that
> time Harvard's Data Science
> Lab won't publish anything I'm afraid I'll need to invent this
> implementation by myself (in which I have less faith :)).
>
>
>
> 2015-09-03 21:46 GMT+02:00 Panos Toulis <notifications@github.com>:
>
> > By the way, if there is still time for your thesis, we will soon be able
> > to share a short document with
> > our implementation of SGD for Cox models :)
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/airoldilab/sgd/issues/23#issuecomment-137555337>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/airoldilab/sgd/issues/23#issuecomment-137685942>.
>



-- 
best
panos

**Comment by MarcinKosinski** [comment]
@ptoulis may I ask one more question:

You wrote:

>
> We are using one of straightforward one where the data are event times
(either something happened or censoring)
Without loss of generality these can be assumed to be ordered.
>

This is rather crucial assumption and it's not so clear and trivial for me. Could you please explain more or post relevant articles :)?

**Comment by ptoulis** [comment]
Take a look at Davison's book. You can find the more detailed argument
there.

On Sat, Sep 5, 2015 at 9:17 AM, Marcin Kosiński <notifications@github.com>
wrote:

> @ptoulis <https://github.com/ptoulis> may I ask one more question:
>
> You wrote:
>
> We are using one of straightforward one where the data are event times
> (either something happened or censoring)
> Without loss of generality these can be assumed to be ordered.
>
> This is rather crucial assumption and it's not so clear and trivial for
> me. Could you please explain more or post relevant articles :)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/airoldilab/sgd/issues/23#issuecomment-137955353>.
>



-- 
best
panos

**Comment by MarcinKosinski** [comment]
It has over 700 pages :) 
I think it is possible to order indexes of observations but they should be randomly picked up for the sgd estimation. That's how I'm planning to do this in my master thesis :)

**Comment by ptoulis** [comment]
I think it mentioned above, that should be Chapter 10 (at least in the 2003
ed.).

On Sun, Sep 6, 2015 at 11:50 AM, Marcin Kosiński <notifications@github.com>
wrote:

> It has over 700 pages :)
> I think it is possible to order indexes of observations but they should be
> randomly picked up for the sgd estimation. That's how I'm planning to do
> this in my master thesis :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/airoldilab/sgd/issues/23#issuecomment-138094586>.
>



-- 
best
panos

**Comment by mikegorczyca** [comment]
I’m a little confused when you say that logistic regression and Cox regression can be non-convex procedures – under standard ML and partial likelihood procedures for logistic regression and Cox regression, respectively, these appear to be convex. Are you referring to a non-standard procedure, such as using a squared loss function for logistic regression (an example is provided below in case I was unclear)? 

> https://math.stackexchange.com/questions/2381724/logistic-regression-when-can-the-cost-function-be-non-convex

If not, could you provide an example where Cox regression under the standard partial likelihood procedure is non-convex? This would be greatly appreciated - I've only been able to generate examples where the Hessian is at least positive semi-definite.

**Comment by ptoulis** [comment]
Indeed, different loss functions can make the objective non-convex. But even with classical loss functions "bad data" can lead to non-convex objectives. Consider for example classical logistic regression where we only observe y=1. The likelihood is  e^θ/ (1 + e^θ), which is non-convex, and is maximized at θ=Inf.  These cases are not frequent in practice but they do exist (e.g., when modeling rare events).

**Comment by mikegorczyca** [comment]
@ptoulis Thank you - this is a good example. I'm assuming you're referring to separation for rare events.

> https://www.jstor.org/stable/2336390?seq=1#page_scan_tab_contents

**Comment by ptoulis** [comment]
Yes, exactly. This is actually a very good reference that we could use.
